### ISSUE TITLE
Question about [bm_tuning of MaxEnt] - [where are the MESS results?]

### ISSUE THREAD

**Issue by zali37** [biomodhub/biomod2#695]
_Please make sure to close the issue once you consider it as solved_
_Please use screenshots only when you cannot copy-paste the object, e.g. for figures or maps_

**Context and question**
Hello Biomod team. I built a model using the MaxEnt algorithm, among others, and selected the MESS option; however, I am unsure how to retrieve the results of that analysis. Could you help me? 
**Related code** _(optional)_

user.maxent <- list("for_all_datasets"=list(
  path_to_maxent.jar="C:/DATOS/ENM/diplolaemus",
  background_data_dir="C:/DATOS/ENM/maxent",
  jackknife=T,
  writeclampgrid=T,
  writemess=T))
                       
opt.dMax <- bm_ModelingOptions(data.type = 'binary',
                            models = "MAXENT",
                            strategy = "user.defined",
                            bm.format = bibronii_format,
                            calib.lines=cv.k,
                            user.val = list("MAXENT.binary.MAXENT.MAXENT"=user.maxent))

tuned.maxent <- bm_Tuning(model = "MAXENT",
                           tuning.fun = "ENMevaluate",
                           do.formula = F,
                           bm.options = opt.dMax@options$MAXENT.binary.MAXENT.MAXENT,
                           bm.format = bibronii_format,
                           calib.lines = cv.k,
                           metric.eval = "auc.val.avg",
                           params.train = list(
                             MAXENT.algorithm = "maxnet",
                             MAXENT.parallel=F,
                             MAXENT.tune.args = list(fc=c("L","Q","LQ","LQH"),rm=1:5),
                             MAXENT.partitions = "randomkfold",
                             MAXENT.kfolds=3
                           ),
                           ctrl.train = NULL
)

 bibronii_opt <- bm_ModelingOptions(
  data.type = "binary",
  models=c("SRE","RF", "GAM.gam.gam","ANN","GLM","MAXENT"),
  bm.format = bibronii_format,
  strategy = "user.defined",
  user.base = "bigboss",
  calib.lines = cv.k,
  user.val = list(SRE.binary.biomod2.bm_SRE = opt.sre,
                  RF.binary.randomForest.randomForest = tuned.rf,
                  GAM.binary.gam.gam = tuned.gam,
                  ANN.binary.nnet.nnet = tuned.ann,
                  GLM.binary.stats.glm = user.GLM,
                  MAXENT.binary.MAXENT.MAXENT = tuned.maxent)
)
bibronii_modelo <- BIOMOD_Modeling(
  bm.format = bibronii_format,
  modeling.id = "solo_bibronii",
  models = c("SRE","ANN","RF","GAM","GLM","MAXENT"),
  CV.strategy = "user.defined",
  CV.user.table = cv.k,
  var.import = 3,
  OPT.strategy = "user.defined",
  OPT.user = bibronii_opt,
  metric.eval = c("AUCroc","TSS"),
  seed.val = 4
)
```
Thank you so much.
Regards,
Natalia

**Comment by MayaGueguen** [comment]
Hello Natalia,

Have you checked within the modeling folder defined by : 
```
file.path(bibronii_modelo@dir.name, bibronii_modelo@sp.name, "models", bibronii_modelo@modeling.id)
```
? 👀 

Looking for a MAXENT folder, such as `[sp.name]_allData_RUN1_MAXENT_outputs` for example ?

Maya

**Comment by zali37** [comment]
Hello Maya 
Yes, I looked everywhere in the folder, but I didn´t find the mess results. So, I re-run the maxent and found this: 
> tuned.maxent <- bm_Tuning(model = "MAXENT",
+                           tuning.fun = "ENMevaluate",
+                           do.formula = F,
+                           bm.options = opt.dDMax@options$MAXENT.binary.MAXENT.MAXENT,
+                           bm.format = darwinii_format,
+                           calib.lines = cv.kD,
+                           metric.eval = "auc.val.avg",
+                           params.train = list(
+                             MAXENT.algorithm = "maxnet",
+                             MAXENT.parallel=F,
+                             MAXENT.tune.args = list(fc=c("L","H","LQ","LQH"),rm=0.5:3),
+                             MAXENT.partitions = "randomkfold",
+                             MAXENT.kfolds=3
+                           ),
+                           ctrl.train = NULL
+ )

		> Dataset _PA1_RUN1
			> Tuning parameters...
*** Running initial checks... ***

* Variable values were input along with coordinates and not as raster data, so no raster predictions can be generated, and AICc is calculated with background data for Maxent models.
* Model evaluations with random 3-fold cross validation...
Do you know exactly what this means? If the mess wasn't computerised, how can I fix it?  
Best wishes
Natalia

**Comment by MayaGueguen** [comment]
Hello Natalia,

Sorry, I did not pay enough attention... 🙈 

You set the `writemess` modeling option to `TRUE`, yes, but within the `user.maxent` object.
Which you do not use after, because you use the `tuned.maxent` object.
So you are here in a complicated situation where you want both to : 
- use the tuned parameters obtained from the [bm_Tuning](https://biomodhub.github.io/biomod2/reference/bm_Tuning.html) function
- but also modify by hand some of the parameters

The solution would be to : 
1. tune maxent over your data and cross-validation partitions
2. update these modeling options with the few additional parameters you want to fix

It would be something like that : 
```
## For each partition, add the fixed parameter values
user.maxent <- tuned.maxent
for (i in 1:length(user.maxent)) {
   user.maxent[[i]][["path_to_maxent.jar"]] <- "C:/DATOS/ENM/diplolaemus"
   user.maxent[[i]][["background_data_dir"]] <- "C:/DATOS/ENM/maxent"
   user.maxent[[i]][["jackknife"]] <- TRUE
   user.maxent[[i]][["writeclampgrid"]] <- TRUE
   user.maxent[[i]][["writemess"]] <- TRUE
}

## Verify the add by carefully looking at the created object
print(user.maxent, dataset = "_PA1_RUN1")
```

Hope it helps,
Maya
